### DISCUSSION TITLE
Install Postgres Extension manually, then trigger migrate prisma migrate dev.

### DISCUSSION THREAD

**Discussion by MEnnabah** [prisma/prisma#17197] [Q&A]
We use Prisma, Postgres, and postgres_fdw already in our stack. To connect to foreign server, we set the postgres_fdw setup to be as part of migration script of Prisma. Everything works as expected.

Later on, we decided to move the postgres_fdw setup script from the migration script to be triggered manually (in a dev env) or as part of our AWS setup (in a prod env).
Right now, within the development environment, I manually trigger the postgres_fdw setup script in TablePlus, and then I trigger npx prisma migrate dev. However, the following error is thrown:

```
Error: P3006

Migration `20221114103902_init_db` failed to apply cleanly to the shadow database.
Error:
db error: ERROR: server "my_foreign_server_name" does not exist
   0: sql_migration_connector::validate_migrations
           with namespaces=None
             at migration-engine/connectors/sql-migration-connector/src/lib.rs:297
   1: migration_core::state::DevDiagnostic
             at migration-engine/core/src/state.rs:264
```

I honestly don’t know why this is happening. I’m 100% sure the foreign server exists and the extension is installed. However, I do suspect the server doesn’t exists somewhere within the prisma context, probably the shadow database.

Am I doing something wrong? What do I need to do to install the extension and connect to a foreign server as part of a script other than Prisma's migration script?

**Comment by janpio** [comment]
I agree this is probably the shadow database, which does not know how to set up the Postgres extension.

> To connect to foreign server, we set the postgres_fdw setup to be as part of migration script of Prisma. Everything works as expected.
> 
> Later on, we decided to move the postgres_fdw setup script from the migration script to be triggered manually (in a dev env) or as part of our AWS setup (in a prod env).

Can you elaborate on this a bit?

Are you activating and configuring the extension in your migration history anywhere?

**Reply by MEnnabah** [reply]
So initially we used to have this on top of our first (earliest) migration.sql file:

```
CREATE EXTENSION IF NOT EXISTS postgres_fdw;
CREATE SERVER IF NOT EXISTS my_foreign_server_name FOREIGN DATA WRAPPER postgres_fdw OPTIONS (host 'host.docker.internal', port '5430', dbname 'external_db_name', fetch_size '250000');
CREATE USER MAPPING IF NOT EXISTS FOR user SERVER my_foreign_server_name OPTIONS (user 'access_user_name', password 'access_password');
```

The parameters of the foreign server was part of the environment variables and we read them using SQL, the actual code looked very messy. However, it worked very well until we our first deployment to production. The production server prevented statements we use to read environment variables (like `COPY`) and storing them in a temporary table.

So, we decided to run the above commands before we run Prisma migration script. These commands are run manually in case of development environment. So a typical workflow for development environment is I'll:
```
$ docker compose up
```
Then, it will create a container for the app (let's call it `main`) and another container for the database (let's call it `maindb`). Now, I can connect to `maindb` with TablePlus and execute the earlier SQL commands (on production, there's a similar process that executes `postgres_fdw` command as well).

After I execute the above SQL commands, I run:

```
$ docker exec -it main /bin/sh
$ npx prisma migrate dev
```

Executing the migration command is exactly what triggers the error.

We're thinking about the different solutions we have, probably there's a way we can set a pre-migration hook where it's reads an sql file that's in `.gitignore`? Or if this is something not supported by Prisma, maybe we can create a `prisma/migrations/0_local_dev/migration.sql` with the earlier command and add it to `.gitignore`.

I'm glad you're thinking with us about this, thank you very much!

**Reply by janpio** [reply]
Ah, that's a messy situation. As you describe, you can not put your configuration command in `migration.sql` as your server does not allow the commands you need (can you share that version of the SQL please? I see no `COPY` in the variant you shared), and without that being there the servers will of course not be configured in your shadow database.

Do you actually need a configured foreign data wrapper in the shadow database to test the migrations though? If not, you could just put the `CREATE EXTENSION IF NOT EXISTS postgres_fdw;` in the migration file and then later run the server configuration manually after you run `migrate deploy` or `migrate dev` to deploy your migration files to your database.

I think generally this is a use case we have not considered yet - so this is quite exciting. I think we will want to definitely turn this into an issue so we can think about how we can properly allow dynamic configuration of extensions.

**Comment by MEnnabah** [comment]
> Do you actually need a configured foreign data wrapper in the shadow database to test the migrations though?

Yeah we don't really need that in the shadow database, but I reckon that might change as the project evolves. I think what I'll do for now is to have a script that run only in local migrations. So it's consistent for everyone on the team and it's consistent from Prisma side as well.

I didn't really include the COPY command because it's a mess :) but here you go:
```
-- load env vars
DROP TABLE IF EXISTS env;
DROP TABLE IF EXISTS env_tmp;
CREATE TEMP TABLE env_tmp(e text);
CREATE TEMP TABLE env(k text, v text);

COPY env_tmp ( e ) FROM PROGRAM 'env';

INSERT INTO env
SELECT (regexp_split_to_array(e,'={1,1}'))[1],
  (regexp_match(e, '={1,1}(.*)'))[1]
FROM env_tmp;

SELECT * FROM env;

DO $$ 
<<block>>
DECLARE
	external_db_username text := '';
	external_db_password text := '';
	external_db_host text := '';
	external_db_name text := '';
	external_db_port text := '';
	external_db_server_name text := 'my_foreign_server_name';
	current_db_username text := '';
BEGIN
	SELECT v into external_db_username FROM env WHERE k = 'EXTERNAL_DB_USERNAME';
	SELECT v into external_db_password FROM env WHERE k = 'EXTERNAL_DB_PASSWORD';
	SELECT v into external_db_host FROM env WHERE k = 'EXTERNAL_DB_HOST';
	SELECT v into external_db_name FROM env WHERE k = 'EXTERNAL_DB';
	SELECT v into external_db_port FROM env WHERE k = 'EXTERNAL_DB_PORT';
	SELECT v into current_db_username FROM env WHERE k = 'POSTGRES_USER';
	
	DROP SERVER IF EXISTS my_foreign_server_name CASCADE;

	EXECUTE FORMAT('CREATE SERVER IF NOT EXISTS %s FOREIGN DATA WRAPPER postgres_fdw OPTIONS (host %L, port %L, dbname %L, fetch_size %L)', my_foreign_server_name, external_db_host, external_db_port, external_db_server_name, '250000');
	EXECUTE FORMAT('CREATE USER MAPPING IF NOT EXISTS FOR %s SERVER %s OPTIONS (user %L,password %L)', current_db_username, external_db_server_name, external_db_username, external_db_password); 


-- Sorry :)
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
